// jovo-cms-googlesheets/dist/src/GoogleSheetsCMS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jovo_core_1 = require("jovo-core");
const _merge = require("lodash.merge");
const { google, JWT } = require('googleapis');
const util = require("util");
const https = require("https");
const fs = require("fs");
const readdir = util.promisify(fs.readdir);
const readFile = util.promisify(fs.readFile);
const exists = util.promisify(fs.exists);
const KeyValueSheet_1 = require("./KeyValueSheet");
const ResponsesSheet_1 = require("./ResponsesSheet");
const DefaultSheet_1 = require("./DefaultSheet");
const ObjectArraySheet_1 = require("./ObjectArraySheet");
class GoogleSheetsCMS extends jovo_core_1.BaseCmsPlugin {
    constructor(config) {
        super(config);
        this.config = {
            enabled: true,
            credentialsFile: './credentials.json',
            spreadsheetId: undefined,
            access: 'private',
            sheets: [],
        };
        if (config) {
            this.config = _merge(this.config, config);
        }
        this.actionSet = new jovo_core_1.ActionSet([
            'retrieve',
        ], this);
    }
    async install(app) {
        super.install(app);
        app.middleware('setup').use(this.retrieveSpreadsheetData.bind(this));
        const defaultSheetMap = {
            'keyvalue': KeyValueSheet_1.KeyValueSheet,
            'responses': ResponsesSheet_1.ResponsesSheet,
            'objectarray': ObjectArraySheet_1.ObjectArraySheet,
            'default': DefaultSheet_1.DefaultSheet
        };
        if (this.config.sheets) {
            this.config.sheets.forEach((sheet) => {
                let type = undefined;
                if (!sheet.type) {
                    type = 'Default';
                }
                if (sheet.type && defaultSheetMap[sheet.type.toLowerCase()]) {
                    type = sheet.type.toLowerCase();
                }
                if (type) {
                    this.use(new defaultSheetMap[type.toLowerCase()](sheet));
                }
            });
        }
    }
    uninstall(app) {
    }
    async retrieveSpreadsheetData(handleRequest) {
        try {
            if (this.config.credentialsFile) {
                const credentialsFileExists = await exists(this.config.credentialsFile);
                if (credentialsFileExists) {
                    this.jwtClient = await this.initializeJWT();
                    this.jwtClient = await this.authorizeJWT(this.jwtClient);
                }
            }
            await this.middleware('retrieve').run(handleRequest, true);
        }
        catch (e) {
            if (e.constructor.name === 'JovoError') {
                throw e;
            }
            throw new jovo_core_1.JovoError(e.message, jovo_core_1.ErrorCode.ERR_PLUGIN, 'jovo-cms-googlesheets', undefined, undefined, 'https://www.jovo.tech/docs/cms/google-sheets');
        }
    }
    async loadPublicSpreadSheetData(spreadsheetId, sheetPosition = 1) {
        const url = `https://spreadsheets.google.com/feeds/list/${spreadsheetId}/${sheetPosition}/public/values?alt=json`;
        jovo_core_1.Log.verbose('Accessing public spreadsheet: ' + url);
        return await this.getJSON(url);
    }
    loadPrivateSpreadsheetData(spreadsheetId, sheet, range) {
        return new Promise((resolve, reject) => {
            try {
                const sheets = google.sheets('v4');
                sheets.spreadsheets.values.get({
                    auth: this.jwtClient,
                    spreadsheetId,
                    range: `${sheet}!${range}`,
                }, (error, response) => {
                    if (error) {
                        return reject(error);
                    }
                    resolve(response.data.values);
                });
            }
            catch (error) {
                reject(error);
            }
        });
    }
    initializeJWT() {
        return new Promise(async (resolve, reject) => {
            try {
                if (!this.config.credentialsFile) {
                    return reject(new Error('Credentials file is mandatory'));
                }
                const credentialsFileExists = await exists(this.config.credentialsFile);
                if (!credentialsFileExists) {
                    return reject(new Error(`Credentials file doesn't exist`));
                }
                const results = await readFile(this.config.credentialsFile);
                const keyFileObject = JSON.parse(results.toString());
                const jwtClient = new google.auth.JWT(keyFileObject.client_email, null, keyFileObject.private_key, ['https://www.googleapis.com/auth/spreadsheets'], null);
                resolve(jwtClient);
            }
            catch (error) {
                reject(error);
            }
        });
    }
    authorizeJWT(jwtClient) {
        return new Promise((resolve, reject) => {
            try {
                jwtClient.authorize((errorJwt) => {
                    if (errorJwt) {
                        return reject(errorJwt);
                    }
                    resolve(jwtClient);
                });
            }
            catch (error) {
                reject(error);
            }
        });
    }
    getJSON(url) {
        return new Promise((resolve, reject) => {
            https.get(url, (res) => {
                let body = '';
                res.on('data', (chunk) => {
                    body += chunk;
                });
                res.on('end', () => {
                    try {
                        if (res.statusCode === 302) {
                            return reject(new jovo_core_1.JovoError('HTTP Response code 302', jovo_core_1.ErrorCode.ERR_PLUGIN, 'jovo-cms-spreadsheets', undefined, "This might occur when you try to access a private spreadsheet without the permission."));
                        }
                        if (res.statusCode === 400) {
                            return reject(new jovo_core_1.JovoError('HTTP Response code 400', jovo_core_1.ErrorCode.ERR_PLUGIN, 'jovo-cms-spreadsheets', undefined, "This might occur when you use the wrong spreadsheet id."));
                        }
                        resolve(JSON.parse(body));
                    }
                    catch (e) {
                        reject(e);
                    }
                });
            }).on('error', (e) => {
                reject(e);
            });
        });
    }
}
exports.GoogleSheetsCMS = GoogleSheetsCMS;
