// jovo-framework/dist/src/hosts/AzureFunction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AzureFunction {
    constructor(context, req) {
        this.hasWriteFileAccess = false;
        this.req = req;
        this.context = context;
        this.headers = req.headers;
        this.$request = req.body;
    }
    getRequestObject() {
        return this.$request;
    }
    setResponse(obj) {
        return new Promise((resolve) => {
            this.context.res = {
                headers: {
                    'Content-Type': 'application/json',
                },
                statusCode: 200,
                body: obj,
            };
            this.context.done();
            resolve();
        });
    }
    fail(error) {
        if (!this.context.res.statusCode) {
            const responseObj = {
                code: 500,
                msg: error.message,
            };
            if (process.env.NODE_ENV === 'production') {
                responseObj.stack = error.stack;
            }
            this.context.res = {
                headers: {
                    'Content-Type': 'application/json',
                },
                statusCode: 500,
                body: responseObj
            };
            this.context.done();
        }
    }
}
exports.AzureFunction = AzureFunction;
