// jovo-platform-googleassistant/dist/src/core/GoogleActionUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jovo_core_1 = require("jovo-core");
const _get = require("lodash.get");
const uuidv4 = require("uuid/v4");
class GoogleActionUser extends jovo_core_1.User {
    constructor(googleAction) {
        super(googleAction);
        this.$storage = {};
        this.googleAction = googleAction;
        try {
            this.$storage = JSON.parse(_get(googleAction.$originalRequest || googleAction.$request, 'user.userStorage'));
        }
        catch (e) {
        }
    }
    getAccessToken() {
        return this.googleAction.$request.getAccessToken();
    }
    getId() {
        const userId = this.$storage.userId ||
            this.googleAction.$request.getUserId() ||
            uuidv4();
        this.$storage.userId = userId;
        return userId;
    }
    getProfile() {
        return _get(this.googleAction.$originalRequest, 'user.profile');
    }
    getPermissions() {
        return _get(this.googleAction.$originalRequest, 'user.permissions');
    }
    hasPermission(permission) {
        const permissions = this.getPermissions();
        if (!permissions) {
            return false;
        }
        return permissions.includes(permission);
    }
    hasNamePermission() {
        return this.hasPermission('NAME');
    }
    hasPreciseLocationPermission() {
        return this.hasPermission('DEVICE_PRECISE_LOCATION');
    }
    hasCoarseLocationPermission() {
        return this.hasPermission('DEVICE_COARSE_LOCATION');
    }
}
exports.GoogleActionUser = GoogleActionUser;
