// jovo-core/dist/src/SpeechBuilder.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const _sample = require('lodash.sample');
/** Class SpeechBuilder */
class SpeechBuilder {
    /**
     * Constructor
     * @param {Jovo=} jovo instance
     * @public
     */
    constructor(jovo) {
        this.speech = '';
        this.jovo = jovo;
    }
    /**
     * Adds text surrounded by <s> tags
     * @param {string} text
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addSentence(text, condition, probability) {
        if (Array.isArray(text)) {
            return this.addText('<s>' + _sample(text) + '</s>', condition, probability);
        }
        return this.addText('<s>' + text + '</s>', condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as cardinal
     * @param {string} n
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addSayAsCardinal(n, condition, probability) {
        if (Array.isArray(n)) {
            return this.addText('<say-as interpret-as="cardinal">' + _sample(n) + '</say-as>', condition, probability);
        }
        return this.addText('<say-as interpret-as="cardinal">' + n + '</say-as>', condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as ordinal
     * @param {string} n
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addCardinal(n, condition, probability) {
        return this.addSayAsCardinal(n, condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as ordinal
     * @param {string} n
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addSayAsOrdinal(n, condition, probability) {
        if (Array.isArray(n)) {
            return this.addText('<say-as interpret-as="ordinal">' + _sample(n) + '</say-as>', condition, probability);
        }
        return this.addText('<say-as interpret-as="ordinal">' + n + '</say-as>', condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as ordinal
     * @param {string} n
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addOrdinal(n, condition, probability) {
        return this.addSayAsOrdinal(n, condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as characters
     * @param {string} characters
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addSayAsCharacters(characters, condition, probability) {
        if (Array.isArray(characters)) {
            return this.addText('<say-as interpret-as="characters">' + _sample(characters) + '</say-as>', condition, probability);
        }
        return this.addText('<say-as interpret-as="characters">' + characters + '</say-as>', condition, probability);
    }
    /**
     * Adds <say-as> tags with interpret-as characters
     * @param {string} characters
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addCharacters(characters, condition, probability) {
        return this.addSayAsCharacters(characters, condition, probability);
    }
    /**
     * Adds break tag to speech obj
     * @public
     * @param {string} time timespan like 3s, 500ms etc
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addBreak(time, condition, probability) {
        const strengthValues = ['none', 'x-weak', 'weak', 'medium', 'strong', 'x-strong'];
        const breakTime = Array.isArray(time) ? _sample(time) : time;
        if (strengthValues.indexOf(breakTime) > -1) {
            return this.addText('<break strength="' + breakTime + '"/>', condition, probability);
        }
        else {
            return this.addText('<break time="' + breakTime + '"/>', condition, probability);
        }
    }
    /**
     * Adds text to speech
     * @public
     * @param {string} text
     * @param {boolean} condition
     * @param {number} probability
     * @return {SpeechBuilder}
     */
    addText(text, condition, probability) {
        if (typeof condition === "boolean" && condition === false) {
            return this;
        }
        if (typeof probability === "number") {
            if (Math.random() >= probability) {
                return this;
            }
        }
        if (this.speech.length > 0) {
            this.speech += ' ';
        }
        this.speech += Array.isArray(text) ? _sample(text) : text;
        return this;
    }
    /**
     * Adds phoneme tag to speech
     * @public
     * @param {string} text
     * @param {string} ph
     * @param {string} alphabet
     * @return {SpeechBuilder}
     */
    addPhoneme(text, ph, alphabet = 'ipa') {
        return this.addText(`<phoneme alphabet="${alphabet}" ph="${SpeechBuilder.escapeXml(ph)}">${text}</phoneme>`);
    }
    /**
     * Adds an x-sampa phoneme to speech
     * @public
     * @param {string} text
     * @param {string} ph
     * @return {SpeechBuilder}
     */
    addXSampa(text, ph) {
        return this.addPhoneme(text, ph, 'x-sampa');
    }
    /**
     * Adds an ipa phoneme to speech
     * @public
     * @param {string} text
     * @param {string} ph
     * @return {SpeechBuilder}
     */
    addIpa(text, ph) {
        return this.addPhoneme(text, ph, 'ipa');
    }
    /**
     * Returns speech object string
     * @deprecated use toString()
     * @public
     * @return {string}
     */
    build() {
        return this.speech;
    }
    /**
     * Returns SpeechBuilder as a string
     * @return {string}
     */
    toString() {
        return this.speech;
    }
    static toSSML(text) {
        text = text.replace(/<speak>/g, '').replace(/<\/speak>/g, '');
        text = '<speak>' + text + '</speak>';
        if (SpeechBuilder.ESCAPE_AMPERSAND) { // workaround (v1 compatibility)
            text = text.replace(/&/g, 'and');
        }
        return text;
    }
    static removeSSML(ssml) {
        let noSSMLText = ssml.replace(/<speak>/g, '').replace(/<\/speak>/g, '');
        noSSMLText = noSSMLText.replace(/<[^>]*>/g, '');
        return noSSMLText;
    }
    static removeSpeakTags(ssml) {
        return ssml.replace(/<speak>/g, '').replace(/<\/speak>/g, '');
    }
    /**
     * Escapes XML in SSML
     *
     * @see https://stackoverflow.com/questions/7918868/how-to-escape-xml-entities-in-javascript
     */
    static escapeXml(unsafe) {
        return unsafe.replace(/[<>&'"]/g, (c) => {
            switch (c) {
                case '<': return '&lt;';
                case '>': return '&gt;';
                case '&': return '&amp;';
                case '\'': return '&apos;';
                default:
                case '"': return '&quot;';
            }
        });
    }
}
SpeechBuilder.ESCAPE_AMPERSAND = true;
exports.SpeechBuilder = SpeechBuilder;
