// jovo-cms-i18next/dist/src/I18Next.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jovo_core_1 = require("jovo-core");
const _merge = require("lodash.merge");
const fs = require("fs");
const util = require("util");
const path = require("path");
const i18n = require('i18next');
class I18Next extends jovo_core_1.BaseCmsPlugin {
    constructor(config) {
        super(config);
        this.config = {
            filesDir: './i18n',
            // i18next
            load: 'all',
            returnObjects: true,
            interpolation: {
                escapeValue: false,
            },
            resources: undefined,
            debug: false,
            partialBundledLanguages: false,
            lng: undefined,
            fallbackLng: 'dev',
            whitelist: false,
            nonExplicitWhitelist: false,
            preload: false,
            lowerCaseLng: false,
            ns: 'translation',
            defaultNS: 'translation',
            fallbackNS: false,
            saveMissing: false,
            updateMissing: false,
            saveMissingTo: 'fallback',
            missingKeyHandler: false,
            parseMissingKeyHandler: undefined,
            appendNamespaceToMissingKey: false,
            missingInterpolationHandler: undefined,
            simplifyPluralSuffix: true,
            postProcess: false,
            returnNull: true,
            returnEmptyString: true,
            returnedObjectHandler: undefined,
            joinArrays: false,
            overloadTranslationOptionHandler: undefined,
            detection: undefined,
            backend: undefined,
            cache: undefined,
            i18nFormat: undefined,
            initImmediate: true,
            keySeparator: '.',
            nsSeparator: ':',
            pluralSeparator: '_',
            contextSeparator: '_',
            appendNamespaceToCIMode: false,
            compatibilityJSON: 'v3',
        };
        if (config) {
            this.config = _merge(this.config, config);
        }
    }
    install(app) {
        super.install(app);
        app.middleware('setup').use(this.loadFiles.bind(this));
        jovo_core_1.Jovo.prototype.t = function () {
            this.$app.$cms.I18Next.i18n.changeLanguage(this.$request.getLocale());
            return this.$app.$cms.I18Next.i18n.t.apply(this.$app.$cms.I18Next.i18n, arguments);
        };
        jovo_core_1.SpeechBuilder.prototype.t = function () {
            this.jovo.$app.$cms.I18Next.i18n.changeLanguage(this.jovo.$request.getLocale());
            const translatedText = this.jovo.$app.$cms.I18Next.i18n.t.apply(this.jovo.$app.$cms.I18Next.i18n, arguments);
            this.addText(translatedText);
            return this;
        };
        jovo_core_1.SpeechBuilder.prototype.addT = function () {
            this.jovo.$app.$cms.I18Next.i18n.changeLanguage(this.jovo.$request.getLocale());
            const translatedText = this.jovo.$app.$cms.I18Next.i18n.t.apply(this.jovo.$app.$cms.I18Next.i18n, arguments);
            this.addText(translatedText);
            return this;
        };
        jovo_core_1.Cms.prototype.t = function () {
            if (!this.$jovo) {
                return;
            }
            this.$jovo.$app.$cms.I18Next.i18n.changeLanguage(this.$jovo.$request.getLocale());
            return this.$jovo.$app.$cms.I18Next.i18n.t.apply(this.$jovo.$app.$cms.I18Next.i18n, arguments);
        };
    }
    async loadFiles(handleRequest) {
        const readdir = util.promisify(fs.readdir);
        handleRequest.app.$cms.I18Next = {};
        handleRequest.app.$cms.I18Next.resources = {};
        const filesDir = this.config.filesDir || '';
        if (fs.existsSync(filesDir)) {
            const dir = await readdir(filesDir);
            jovo_core_1.Log.verbose(`Iterating i18n folder: ${filesDir}`);
            dir.forEach((file) => {
                const ext = file.split('.')[1];
                const validExtensions = ['js', 'json'];
                if (validExtensions.includes(ext)) {
                    const locale = file.split('.')[0];
                    jovo_core_1.Log.verbose(`- ${file}`);
                    handleRequest.app.$cms.I18Next.resources[locale] = require(path.join(process.cwd(), filesDir, file));
                }
            });
        }
        else if (this.config.resources) {
            handleRequest.app.$cms.I18Next.resources = this.config.resources;
        }
        jovo_core_1.Log.debug(`Adding resources to $cms object:`);
        jovo_core_1.Log.debug(JSON.stringify(handleRequest.app.$cms.I18Next.resources, null, '\t'));
        i18n
            .init(_merge({
            resources: handleRequest.app.$cms.I18Next.resources
        }, this.config));
        handleRequest.app.$cms.I18Next.i18n = i18n;
    }
}
exports.I18Next = I18Next;
