// jovo-platform-googleassistant/dist/src/core/GoogleActionResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jovo_core_1 = require("jovo-core");
const _get = require("lodash.get");
class GoogleActionResponse {
    getSessionData(path) {
        return undefined;
    }
    hasSessionData(name, value) {
        return this.hasSessionAttribute(name, value);
    }
    setSessionData(sessionData) {
        return this;
    }
    getSpeech() {
        if (!_get(this, 'richResponse.items[0].simpleResponse.ssml')) {
            return;
        }
        return jovo_core_1.SpeechBuilder.removeSpeakTags(_get(this, 'richResponse.items[0].simpleResponse.ssml'));
    }
    getReprompt() {
        if (!_get(this, 'noInputPrompts[0].ssml')) {
            return;
        }
        return jovo_core_1.SpeechBuilder.removeSpeakTags(_get(this, 'noInputPrompts[0].ssml'));
    }
    getSpeechPlain() {
        const speech = this.getSpeech();
        if (!speech) {
            return;
        }
        return jovo_core_1.SpeechBuilder.removeSSML(speech);
    }
    getRepromptPlain() {
        const reprompt = this.getReprompt();
        if (!reprompt) {
            return;
        }
        return jovo_core_1.SpeechBuilder.removeSSML(reprompt);
    }
    getSessionAttributes() {
        return undefined;
    }
    setSessionAttributes() {
        return this;
    }
    hasSessionAttribute(name, value) {
        return undefined;
    }
    hasState() {
        return undefined;
    }
    hasSessionEnded() {
        return !this.expectUserResponse;
    }
    isTell(speech) {
        if (this.expectUserResponse === true) {
            return false;
        }
        if (speech) {
            if (Array.isArray(speech)) {
                const results = speech.find((text) => {
                    return text === this.getSpeech();
                });
                if (results && results.length === 0) {
                    return false;
                }
            }
            else {
                if (speech.toString() !== this.getSpeech()) {
                    return false;
                }
            }
        }
        return true;
    }
    isAsk(speech, reprompt) {
        if (this.expectUserResponse === false) {
            return false;
        }
        if (speech) {
            if (Array.isArray(speech)) {
                const results = speech.find((text) => {
                    return text === this.getSpeech();
                });
                if (results && results.length === 0) {
                    return false;
                }
            }
            else {
                if (speech.toString() !== this.getSpeech()) {
                    return false;
                }
            }
        }
        if (reprompt) {
            if (Array.isArray(reprompt)) {
                const results = reprompt.find((text) => {
                    return text === this.getReprompt();
                });
                if (results && results.length === 0) {
                    return false;
                }
            }
            else {
                if (reprompt.toString() !== this.getReprompt()) {
                    return false;
                }
            }
        }
        return true;
    }
    toJSON() {
        return Object.assign({}, this);
    }
    static fromJSON(json) {
        if (typeof json === 'string') {
            return JSON.parse(json, GoogleActionResponse.reviver);
        }
        else {
            const alexaResponse = Object.create(GoogleActionResponse.prototype);
            return Object.assign(alexaResponse, json);
        }
    }
    static reviver(key, value) {
        return key === "" ? GoogleActionResponse.fromJSON(value) : value;
    }
}
exports.GoogleActionResponse = GoogleActionResponse;
